xraylog/xray_log.go

```diff
@@ -0,0 +1,77 @@
+// Copyright 2017-2017 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License"). You may not use this file except in compliance with the License. A copy of the License is located at
+//
+//     http://aws.amazon.com/apache2.0/
+//
+// or in the "license" file accompanying this file. This file is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the specific language governing permissions and limitations under the License.
+
+package xraylog
+
+import (
+	"fmt"
+	"io"
+	"time"
+)
+
+// Logger is the logging interface used by xray. fmt.Stringer is used to
+// defer expensive serialization operations until the message is actually
+// logged (i.e. don't bother serializing debug messages if they aren't going
+// to show up).
+type Logger interface {
+	Log(level LogLevel, msg fmt.Stringer)
+}
+
+// LogLevel represents the severity of a log message, where a higher value
+// means more severe. The integer value should not be serialized as it is
+// subject to change.
+type LogLevel int
+
+const (
+	LogLevelDebug LogLevel = iota + 1
+	LogLevelInfo
+	LogLevelWarn
+	LogLevelError
+)
+
+func (ll LogLevel) String() string {
+	switch ll {
+	case LogLevelDebug:
+		return "DEBUG"
+	case LogLevelInfo:
+		return "INFO"
+	case LogLevelWarn:
+		return "WARN"
+	case LogLevelError:
+		return "ERROR"
+	default:
+		return fmt.Sprintf("UNKNOWN<%d>", ll)
```

Comment:
I would suggest by default, the local level should be `INFO` even if the format cannot be recognized.

Reply:
This is just the `String()` of the format, so the return value will not affect what log messages show up. The default log level is already "info" (this method is just for displaying the log level as a string).

It shouldn't be possible in normal circumstances to get an unknown log level. But if it does happen somehow, something weird is happening and logging may not be working as expected,  so this tells the developer/user that something is wrong.

Reply:
Will `UNKNOWNLOGLEVEL` be more informative?

Reply:
Done.